src/query/plan/rewrite/edge_index_lookup.hpp

```diff
@@ -0,0 +1,919 @@
+  }
+
+  bool PostVisit(PeriodicSubquery & /*op*/) override {
+    prev_ops_.pop_back();
+    return true;
+  }
+
+  std::shared_ptr<LogicalOperator> new_root_;
+
+ private:
+  SymbolTable *symbol_table_;
+  AstStorage *ast_storage_;
+  TDbAccessor *db_;
+  // Collected filters, pending for examination if they can be used for advanced
+  // lookup operations (by index, node ID, ...).
+  Filters filters_;
+  // Expressions which no longer need a plain Filter operator.
+  std::unordered_set<Expression *> filter_exprs_for_removal_;
+  std::vector<LogicalOperator *> prev_ops_;
+  std::unordered_set<Symbol> additional_bound_symbols_;
+
+  struct EdgeTypePropertyIndexInfo {
+    std::optional<LabelIx> edge_type_from_filter{};
+    std::optional<storage::EdgeTypeId> edge_type_from_relationship{};
+    // FilterInfo with PropertyFilter.
+    FilterInfo filter;
+    int64_t edge_count;
+  };
+
+  struct CandidateIndex {
+    std::optional<LabelIx> edge_type_from_filter{};
+    std::optional<storage::EdgeTypeId> edge_type_from_relationship{};
+    PropertyIx property;
+    FilterInfo filter;
+  };
+
+  storage::EdgeTypeId GetEdgeType(const EdgeTypeIx &edge_type) { return db_->NameToEdgeType(edge_type.name); }
+  storage::EdgeTypeId GetEdgeType(const LabelIx &edge_type) {
+    return storage::EdgeTypeId::FromUint(db_->NameToLabel(edge_type.name).AsUint());
+  }
+  storage::EdgeTypeId GetEdgeType(const EdgeTypePropertyIndexInfo &info) {
+    return info.edge_type_from_filter.has_value() ? GetEdgeType(info.edge_type_from_filter.value())
+                                                  : info.edge_type_from_relationship.value();
+  }
+  storage::EdgeTypeId GetEdgeType(const CandidateIndex &candidate) {
+    return candidate.edge_type_from_filter.has_value() ? GetEdgeType(candidate.edge_type_from_filter.value())
+                                                       : candidate.edge_type_from_relationship.value();
+  }
+
+  bool FoundIndexWithFilteredLabel(const EdgeTypePropertyIndexInfo &info) {
+    return info.edge_type_from_filter.has_value();
+  }
+
+  EdgeTypePropertyIndexInfo ConstructEdgeTypePropertyIndexInfo(CandidateIndex candidate, int64_t edge_count) {
+    if (candidate.edge_type_from_relationship.has_value()) {
+      return EdgeTypePropertyIndexInfo{.edge_type_from_relationship = candidate.edge_type_from_relationship,
+                                       .filter = candidate.filter,
+                                       .edge_count = edge_count};
+    }
+
+    return EdgeTypePropertyIndexInfo{
+        .edge_type_from_filter = candidate.edge_type_from_filter, .filter = candidate.filter, .edge_count = edge_count};
+  }
+
+  storage::PropertyId GetProperty(const PropertyIx &prop) { return db_->NameToProperty(prop.name); }
+
+  std::vector<CandidateIndex> GetCandidateIndicesFromFilter(const Symbol &symbol) {
+    std::vector<CandidateIndex> candidate_indices{};
+    for (const auto &edge_type : filters_.FilteredLabels(symbol)) {
```

Comment:
Label != EdgeType some semantic/naming inconsistency

Reply:
This one is a bit tricky, as in the filters_ they kinda messed up the logic when extracting these from the relationship (they call it labels), when in fact these are edge types. I'll see how to best introduce this

Reply:
I would keep this out of scope for this PR just because it has been so far semantically incorrect.
e.g.
match (n)-[r:TYPE]->(m) -> this would be translated into an EdgeTypeIx -> since the pattern processor knew at that time that it's an edge
MATCH (n)-[r]->(m) with r where r:TYPE -> this would be translated into a LabelIx as it was processed as a part of the filter

FilteredLabels then returns both label and edge types under LabelIx, regardless of the symbol